### ISSUE TITLE
[java] AccessorMethodGeneration false positives with compile time constants

### ISSUE THREAD

**Issue by TWiStErRob** [pmd/pmd#808]
**Affects PMD Version:** 5.8.1, 6.0.0 (and probably 5.5.4 upwards since the rule was introduced)

**Rule:** `AccessorMethodGeneration`

**Description:** constants are inlined at compile time, so no method is generated to access them.
Valid violations confirmed via `javac Outer.java && javap Outer.class` which shows the `access$` methods.

**Code Sample demonstrating the issue:**

```java
@SuppressWarnings("unused")
class Outer {
  private static final String CONST_STRING = "value";
  private static final String CALCULATED_STRING = "value" + 0;
  private static final int LITERAL = 0;
  private static final int CALCULATED = LITERAL * 2;
  private static final int REFERENCED = Thread.MIN_PRIORITY;
  private static final int CAST = (int) 0L;
  private static final long NON_CONSTANT = java.util.concurrent.TimeUnit.SECONDS.toMillis(10);
  private static final int LATE_INIT;
  static {
  	LATE_INIT = 0;
  }

  class Inner {
    @Override
    public String toString() {
      return "" // separate lines so the rule violations show up better
           + CONST_STRING
           + CALCULATED_STRING
           + LITERAL
           + CALCULATED
           + LATE_INIT    // valid violation
           + REFERENCED
           + CAST
           + NON_CONSTANT // valid violation
      ;
    }
  }
}
```

**Running PMD through:** Gradle (5.8.1) and CLI (6.0.0 `pmd -d . -format text -R java-design`)

### Workaround
I was able to suppress the most basic false positives (CONST_STRING and LITERAL) via:
```xml
<rule ref="rulesets/java/design.xml/AccessorMethodGeneration">
        <properties>
            <!-- Ignore references to `private static final * * = <literal>`
                 Suppress via XPath: current node (access that generates the accessor) is .
                 Check if there exists a FieldDeclaration (private static final)
                 which has a VariableInitializer with a Literal
                 and the name (@Image) of the declaration is the same as the accessed member.
                 TODO calculated constant values still show up as false positive
            -->
            <property name="violationSuppressXPath" value="
            .[@Image =
                //FieldDeclaration[@Private = 'true' and @Static='true' and @Final='true']
                /VariableDeclarator[
                    VariableInitializer/Expression/PrimaryExpression[not(PrimarySuffix)]/PrimaryPrefix/Literal
                ]/VariableDeclaratorId/@Image
             ]" />
        </properties>
    </rule>
```

**Comment by jsotuyod** [comment]
@TWiStErRob very interesting analysis. Definitely a nice avenue of improvement.

**Comment by Shubham-2k17** [comment]
Hi @jsotuyod ,
I would like to solve this issue.
Any hint how on how to get the assigned values(whatever is assigned/ content after =) inside **AccessorMethodGenerationRule**.

**Comment by jsotuyod** [comment]
@Shubham-2k17 really appreciated. Bare in mind this is not going to be as easy as other issues you have taken on so far, but I'll gladly help you along the way either here or on gitter.

The main focus should be on the `analyzeMember ` method:

https://github.com/pmd/pmd/blob/dab1fb24c6fe982d0761d3daebecd4d223697c15/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/AccessorMethodGenerationRule.java#L54-L68

This method is called to analyze, both fields AND methods, making sure that, if they are private, they are only used within the same scope that defines them.

Here, we would have to check **for `ASTFieldDeclaration`s only**, all the child `ASTVariableDeclarator` (`field.findChildrenOfType(ASTVariableDeclarator.class)`), and for each one, check if it has a `ASTVariableInitializer` child (initialization at declaration is optional). If it does, you would have to check the contents of the initializer.

Now, here comes the hardest part. As shown in the original report, we are looking for 2 different things:
1. literals (which are constants)
2. constant expressions (expressions composed exclusively of other literals or constant expressions)

Checking for literals should be somewhat easy. The second part is somewhat harder. I'd recommend you submit a first PR simply addressing the literal part to get started.

The designer shipped with PMD (`./run.sh designer`) should be of great help to better understand the AST generated by the sample code and how to navigate it.

**Comment by TWiStErRob** [comment]
@jsotuyod really like how contributors are treated, lot of help, well done!

re 2., is it possible to find out that a constant used from another class is a compile time literal? in my report it's the `REFERENCED` constant; that constant may be in another library and may or may not be constant.

**Comment by jsotuyod** [comment]
@TWiStErRob yes, doing that is possible through type resolution. Provided a full auxclasspath PMD can do it.

**Comment by Shubham-2k17** [comment]
Hi @jsotuyod,
How to differentiate between referenced and non constant.
```
private static final int REFERENCED = Thread.MIN_PRIORITY;
private static final long NON_CONSTANT = java.util.concurrent.TimeUnit.SECONDS.toMillis(10);
```
I am getting Thread and java.util.concurrent.TimeUnit in prefix which is getting treated similarly. How to differentiate 1st one as referenced constant & 2nd one as function call.
Any hint will help.

**Comment by jsotuyod** [comment]
@Shubham-2k17 the designer is of great help here.

![image](https://user-images.githubusercontent.com/802626/52054807-6c7e6900-253c-11e9-9779-4859e4556771.png)

As you have seen, in both cases you get a `PrimaryPrefix` with the name, but on the method call, there is also a `PrimarySufix` with an `Arguments` and `ArgumentList`. Such suffix is non-existing unless there is a method call. So, looking for a `PrimaryExpression` and checking if it has a `PrimarySufix` or not should suffice.

**Comment by TWiStErRob** [comment]
`private static final int CALCULATED_REFERENCE = Integer.MAX_VALUE / 2;` still yields a violation

**Comment by TWiStErRob** [comment]
Also is `if (isFinal())` enough there? `private final int FOO = 1;` is not a compile time constant and must be reported as a violation; there was no test for this, so it was possible to break it.

**Comment by jsotuyod** [comment]
> `private static final int CALCULATED_REFERENCE = Integer.MAX_VALUE / 2;` still yields a violation

That scenario is beyond what is doable without looking at individual bytecode.

`Integer.MAX_VALUE` is defined outside of the current project, so we don't have source available for symbol table even if we had multifile analysis (we still don't support it). We can look into `Integer.MAX_VALUE` through reflection thanks to type resolution, and we could check if it's static and final, and we may even get it's value, but that is not enough to know if it's a compile-time constant.

Take a look at this example:

```java
public class Test {
	private static final int INITIALIZED = 1;
	private static final int UNINITIALIZED;
	private final int NON_STATIC = 3;

	static {
		UNINITIALIZED = 2;
	}

	public static class Inner {
		public void foo(Test t) {
			System.out.println(Test.INITIALIZED);
			System.out.println(Test.UNINITIALIZED);
			System.out.println(t.NON_STATIC);
		}
	}
}
```

Which ones are actually compile time constants? Well, according to `javap`…

Test.class
```
public class Test {
  public Test();
    descriptor: ()V
    Code:
       0: aload_0
       1: invokespecial #2                  // Method java/lang/Object."<init>":()V
       4: aload_0
       5: iconst_3
       6: putfield      #3                  // Field NON_STATIC:I
       9: return
    LineNumberTable:
      line 1: 0
      line 4: 4

  static int access$000();
    descriptor: ()I
    Code:
       0: getstatic     #1                  // Field UNINITIALIZED:I
       3: ireturn
    LineNumberTable:
      line 1: 0

  static {};
    descriptor: ()V
    Code:
       0: iconst_2
       1: putstatic     #1                  // Field UNINITIALIZED:I
       4: return
    LineNumberTable:
      line 7: 0
      line 8: 4
}
```

Only `UNINITIALIZED` generates an accessor method, as it's the only non-final value not set in an initializer.

Looking into `Test.Inner` confirms this:

```
public class Test$Inner {
  public Test$Inner();
    descriptor: ()V
    Code:
       0: aload_0
       1: invokespecial #2                  // Method java/lang/Object."<init>":()V
       4: return
    LineNumberTable:
      line 10: 0

  public void foo(Test);
    descriptor: (LTest;)V
    Code:
       0: getstatic     #3                  // Field java/lang/System.out:Ljava/io/PrintStream;
       3: iconst_1
       4: invokevirtual #4                  // Method java/io/PrintStream.println:(I)V
       7: getstatic     #3                  // Field java/lang/System.out:Ljava/io/PrintStream;
      10: invokestatic  #5                  // Method Test.access$000:()I
      13: invokevirtual #4                  // Method java/io/PrintStream.println:(I)V
      16: getstatic     #3                  // Field java/lang/System.out:Ljava/io/PrintStream;
      19: iconst_3
      20: invokevirtual #4                  // Method java/io/PrintStream.println:(I)V
      23: return
    LineNumberTable:
      line 12: 0
      line 13: 7
      line 14: 16
      line 15: 23
}
```

Clearly, `iconst_1` and `iconst_3` in place for `INITIALIZED` and `NON_STATIC`.

It's not enough to know a field is final, and it being static has nothing to do with it, but we need to know when and how the field is initialized.

> Also is `if (isFinal())` enough there? `private final int FOO = 1;` is not a compile time constant and must be reported as a violation; there was no test for this, so it was possible to break it.

As per the previous example, if it's final AND it's initialized upon definition, then it's treated as a compile-time constant.

**Comment by TWiStErRob** [comment]
> we could check if it's static and final, and we may even get it's value, but that is not enough to know if it's a compile-time constant.

O.o if it's in the .class file (accessed through aux-classpath) and available through bytecode after compilation without running code, isn't it kind of a "definition" of a compile-time constant?

> and it being static has nothing to do with it

Wow, I did not know non-static can be constant, but indeed:
```java
class Outer {
	private final String field = "asdf";
	class Inner {
		Inner() { System.out.println(field); }
	}
}
```
in this case `Outer() { field = "asdf" }` is generated, but the value is also inlined (`ldc`) into `Inner` constructor, wow. I guess that's a liberty of javac, since we said `final`: even though we can set it through reflection, `Inner` will never observe the new value:
```java
Outer() {
	try {
		java.lang.reflect.Field f = getClass().getDeclaredField("field");
		f.setAccessible(true);
		f.set(this, "foo");
	} catch (Exception ex) { ex.printStackTrace(); }
}
public static void main(String[] args) { new Outer().new Inner(); }
```

Thanks for the thorough explanation.

Re
> Checking for literals should be somewhat easy. The second part is somewhat harder. I'd recommend you submit a first PR simply addressing the literal part to get started.

Is there a tracking issue for the second part?

**Comment by jsotuyod** [comment]
@TWiStErRob you understood perfectly. Java will keep the field, as the "exposed" members remain (even if the visibility makes them unaccessible without reflection), so that developers will always find what they are looking for, but if it's `final` and it knows for sure what the value is at compile-time, the generated bytecode will inline it everywhere to avoid unnecessary load operations.

> Is there a tracking issue for the second part?

There isn't. But it's worth noting, the submited PR does cover a number of constant expressions. In particular, same-class constant expressions should be ok (as there can be no forward-references):

```java
private static final int FOO = 1;
private static final int BAR = FOO * 3;
```

The current code will recognize both as compile-time constants.

For external constants (such as `Integer.MAX_VALUE`), as already stated, there is nothing we can do.

What we are not covering right now are:

1. constants defined in another file of the analyzed project (not doable without multifile analysis)
2. constants defined in another class within the same source file

With the current PMD capabilities, only the second one would be addressable.
